README.md
# whatsapp_camera
<p>
This is a package to open a camera along with a photo gallery, to simplify the steps of the end user
</p>
<p>
It also has a widget for viewing images
</p>

### how to use

Open camera:
```dart
List<File>? res = await Navigator.push(
  context, MaterialPageRoute(
    builder: (context) => const WhatsappCamera(),
  ),
);
```
<br>

Open image: 
```dart
Navigator.push(
  context, MaterialPageRoute(
    builder: (context) => const ViewImage(
      image: 'https://...', 
      imageType: ImageType.network,
    ),
  ),
);
```
<p align="center">
<img  src="https://github.com/welitonsousa/whatsapp_camera/blob/main/assets/example.gif?raw=true" width="250" height="500"/>
</p>

<hr>

## Android
add permissions: <br>
<b>file:</b>  `/android/app/main/AndroidManifest.xml`

```dart
<uses-permission android:name="android.permission.WRITE_EXTERNAL_STORAGE" />
<uses-permission android:name="android.permission.READ_EXTERNAL_STORAGE" />
<application
  android:requestLegacyExternalStorage="true"
  ...
```
<b>file:</b>  `android/app/build.gradle`
```dart
minSdkVersion 21
compileSdkVersion 33
```

## ios
<b>file:</b> `/ios/Runner/Info.plist`

```dart
<key>NSCameraUsageDescription</key>
<string>Can I use the camera please?</string>
<key>NSPhotoLibraryUsageDescription</key>
<string>Used to demonstrate image picker plugin</string>
```

 <a target="_blank" href="https://github.com/welitonsousa/whatsapp_camera/blob/main/LICENSE">LICENSE</a>

<br>
<br>
<p align="center">
   Feito com ❤️ by <a target="_blank" href="https://github.com/welitonsousa"><b>Weliton Sousa</b></a>
</p>


CHANGELOG.md
## 1.0.0

* TODO: Release.


example/lib/main.dart
import 'dart:io';

import 'package:flutter/material.dart';
import 'package:whatsapp_camera/whatsapp_camera.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return const MaterialApp(home: Home());
  }
}

class Home extends StatefulWidget {
  const Home({super.key});

  @override
  State<Home> createState() => _HomeState();
}

class _HomeState extends State<Home> {
  final files = ValueNotifier(<File>[]);

  @override
  void initState() {
    files.addListener(() => setState(() {}));
    super.initState();
  }

  @override
  void dispose() {
    files.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: const Text("whatsapp camera")),
      floatingActionButton: FloatingActionButton(
        child: const Icon(Icons.camera),
        onPressed: () async {
          List<File>? res = await Navigator.push(
            context,
            MaterialPageRoute(
              builder: (context) => const WhatsappCamera(),
            ),
          );
          if (res != null) files.value = res;
        },
      ),
      body: ListView.builder(
        itemCount: files.value.length,
        itemBuilder: (context, index) {
          return SizedBox(
            height: 200,
            width: 200,
            child: Image.file(files.value[index]),
          );
        },
      ),
    );
  }
}


lib/camera/camera.dart
import 'package:flutter/gestures.dart';
import 'package:flutter/material.dart';
import 'package:flutter_sliding_up_panel/flutter_sliding_up_panel.dart';
import 'package:photo_gallery/photo_gallery.dart';
import 'package:whatsapp_camera/camera/camera_whatsapp_controller.dart';
import 'package:whatsapp_camera/camera/images_page.dart';

class WhatsappCamera extends StatefulWidget {
  final bool multiple;
  const WhatsappCamera({super.key, this.multiple = true});

  @override
  State<WhatsappCamera> createState() => _WhatsappCameraState();
}

class _WhatsappCameraState extends State<WhatsappCamera> {
  late WhatsAppCameraController controller;
  final painel = SlidingUpPanelController();

  @override
  void dispose() {
    controller.dispose();
    painel.dispose();
    super.dispose();
  }

  @override
  void initState() {
    controller = WhatsAppCameraController(multiple: widget.multiple);
    controller.inicialize();
    painel.addListener(() {
      if (painel.status.name == 'hidden') {
        controller.selectedImages.clear();
      }
    });
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: Colors.black,
      body: Stack(
        children: [
          SizedBox(
            width: MediaQuery.of(context).size.width,
            // child: CameraCamera(
            //   enableZoom: false,
            //   resolutionPreset: ResolutionPreset.high,
            //   onFile: (file) {
            //     controller.captureImage(file);
            //     Navigator.pop(context, controller.selectedImages);
            //   },
            // ),
          ),
          Padding(
            padding: const EdgeInsets.only(top: 30),
            child: Row(
              mainAxisAlignment: MainAxisAlignment.spaceBetween,
              children: [
                IconButton(
                  color: Colors.white,
                  onPressed: (() => Navigator.pop(context)),
                  icon: const Icon(Icons.close),
                ),
                IconButton(
                  color: Colors.white,
                  onPressed: () async {
                    controller.openGallery().then((value) {
                      Navigator.pop(context, controller.selectedImages);
                    });
                  },
                  icon: const Icon(Icons.image),
                ),
              ],
            ),
          ),
          Positioned(
            bottom: 96,
            left: 0.0,
            right: 0.0,
            child: GestureDetector(
              dragStartBehavior: DragStartBehavior.down,
              onVerticalDragStart: (details) => painel.expand(),
              child: SizedBox(
                height: 120,
                child: AnimatedBuilder(
                    animation: controller,
                    builder: (context, child) {
                      return Column(
                        children: [
                          if (controller.images.isNotEmpty)
                            const RotatedBox(
                              quarterTurns: 1,
                              child: Icon(
                                Icons.arrow_back_ios,
                                color: Colors.white,
                              ),
                            ),
                          Expanded(
                            child: ListView.builder(
                              itemCount: controller.images.length,
                              physics: const BouncingScrollPhysics(),
                              scrollDirection: Axis.horizontal,
                              itemBuilder: (context, index) {
                                return InkWell(
                                  onTap: () async {
                                    controller
                                        .selectImage(controller.images[index])
                                        .then((value) {
                                      Navigator.pop(
                                        context,
                                        controller.selectedImages,
                                      );
                                    });
                                  },
                                  child: Container(
                                    height: 100,
                                    width: 100,
                                    margin: const EdgeInsets.symmetric(
                                        horizontal: 5),
                                    decoration: BoxDecoration(
                                      image: DecorationImage(
                                        fit: BoxFit.cover,
                                        isAntiAlias: true,
                                        filterQuality: FilterQuality.high,
                                        image: ThumbnailProvider(
                                          highQuality: true,
                                          mediumId: controller.images[index].id,
                                        ),
                                      ),
                                    ),
                                  ),
                                );
                              },
                            ),
                          ),
                        ],
                      );
                    }),
              ),
            ),
          ),
          Center(
            child: SlidingUpPanelWidget(
              controlHeight: 0,
              panelController: painel,
              child: AnimatedBuilder(
                  animation: controller,
                  builder: (context, child) {
                    return ImagesPage(
                      controller: controller,
                      close: () {
                        painel.hide();
                      },
                      done: () {
                        if (controller.selectedImages.isNotEmpty) {
                          Navigator.pop(context, controller.selectedImages);
                        } else {
                          painel.hide();
                        }
                      },
                    );
                  }),
            ),
          )
        ],
      ),
    );
  }
}


lib/camera/view_image.dart
import 'dart:io';
import 'package:flutter/material.dart';
import 'package:photo_view/photo_view.dart';

enum ImageType { file, network, asset }

class ViewImage extends StatefulWidget {
  ///
  /// image path: network url, file path or asset
  ///
  final String image;

  ///
  /// define the image type
  ///
  final ImageType imageType;
  const ViewImage({
    super.key,
    required this.image,
    this.imageType = ImageType.file,
  });

  @override
  State<ViewImage> createState() => _ViewImageState();
}

class _ViewImageState extends State<ViewImage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        backgroundColor: Colors.black,
        elevation: 0,
      ),
      backgroundColor: Colors.black,
      body: PhotoView(
        enablePanAlways: true,
        imageProvider: imageProvider,
        maxScale: 3.0,
      ),
    );
  }

  ImageProvider get imageProvider {
    if (widget.imageType == ImageType.file) {
      return FileImage(File(widget.image));
    } else if (widget.imageType == ImageType.network) {
      return NetworkImage(widget.image);
    } else {
      return AssetImage(widget.image);
    }
  }
}


lib/camera/images_page.dart
import 'package:flutter/material.dart';
import 'package:photo_gallery/photo_gallery.dart';
import 'package:whatsapp_camera/camera/camera_whatsapp_controller.dart';
import 'package:whatsapp_camera/camera/view_image.dart';

class ImagesPage extends StatefulWidget {
  final WhatsAppCameraController controller;
  final void Function()? close;
  final void Function()? done;
  const ImagesPage({
    super.key,
    required this.controller,
    this.close,
    this.done,
  });

  @override
  State<ImagesPage> createState() => _ImagesPageState();
}

class _ImagesPageState extends State<ImagesPage> {
  @override
  Widget build(BuildContext context) {
    return Container(
      color: Colors.black,
      height: MediaQuery.of(context).size.height - 40,
      child: Column(
        children: [
          Container(
            height: 40,
            decoration: const BoxDecoration(
                color: Colors.white,
                borderRadius: BorderRadius.only(
                  topLeft: Radius.circular(15),
                  topRight: Radius.circular(15),
                )),
            child: Row(
              mainAxisAlignment: MainAxisAlignment.spaceBetween,
              children: [
                IconButton(
                  onPressed: widget.close?.call,
                  icon: const Icon(Icons.close),
                ),
                if (widget.controller.multiple)
                  Text(widget.controller.selectedImages.length.toString()),
                IconButton(
                  onPressed: widget.done?.call,
                  icon: const Icon(Icons.check),
                )
              ],
            ),
          ),
          Expanded(
            child: GridView.builder(
              padding: const EdgeInsets.symmetric(horizontal: 5),
              itemCount: widget.controller.images.length,
              physics: const BouncingScrollPhysics(),
              gridDelegate: SliverGridDelegateWithFixedCrossAxisCount(
                crossAxisCount: 3,
                mainAxisExtent: 140,
                crossAxisSpacing: 5,
                mainAxisSpacing: 5,
                childAspectRatio: MediaQuery.of(context).size.width /
                    (MediaQuery.of(context).size.height / 4),
              ),
              itemBuilder: (context, index) {
                return InkWell(
                  onTap: () => widget.controller
                      .selectImage(widget.controller.images[index]),
                  child: ImageItem(
                    selected: widget.controller.imageIsSelected(
                      widget.controller.images[index].filename,
                    ),
                    image: widget.controller.images[index],
                  ),
                );
              },
            ),
          ),
        ],
      ),
    );
  }
}

class ImageItem extends StatelessWidget {
  final Medium image;
  final bool selected;
  const ImageItem({super.key, required this.image, this.selected = false});

  @override
  Widget build(BuildContext context) {
    return InkWell(
      child: Stack(
        children: [
          Hero(
            tag: image.id,
            child: Container(
              decoration: BoxDecoration(
                image: DecorationImage(
                  fit: BoxFit.cover,
                  filterQuality: FilterQuality.high,
                  image: ThumbnailProvider(
                    mediumId: image.id,
                    highQuality: true,
                    height: 150,
                    width: 150,
                    mediumType: MediumType.image,
                  ),
                ),
              ),
            ),
          ),
          if (selected)
            Container(
              color: Colors.grey.withOpacity(.3),
              child: Center(
                child: Stack(
                  children: [
                    const Icon(
                      Icons.done,
                      size: 52,
                      color: Colors.white,
                    ),
                    Icon(
                      Icons.done,
                      size: 50,
                      color: Theme.of(context).primaryColor,
                    ),
                  ],
                ),
              ),
            ),
          Align(
            alignment: Alignment.topRight,
            child: IconButton(
              color: Colors.white,
              icon: const Icon(Icons.zoom_out_map_outlined),
              onPressed: () async {
                image.getFile().then((value) {
                  Navigator.of(context).push(MaterialPageRoute(
                    builder: (context) {
                      return Hero(
                        tag: image.id,
                        child: ViewImage(image: value.path),
                      );
                    },
                  ));
                });
              },
            ),
          )
        ],
      ),
    );
  }
}


lib/camera/camera_whatsapp_controller.dart
import 'dart:io';
import 'package:file_picker/file_picker.dart';
import 'package:flutter/material.dart';
import 'package:photo_gallery/photo_gallery.dart';

class WhatsAppCameraController extends ChangeNotifier {
  WhatsAppCameraController({this.multiple = false});
  final bool multiple;
  final selectedImages = <File>[];
  var images = <Medium>[];

  bool imageIsSelected(String? fileName) {
    final index =
        selectedImages.indexWhere((e) => e.path.split('/').last == fileName);
    return index != -1;
  }

  Future<void> inicialize() async {
    final albums = await PhotoGallery.listAlbums(mediumType: MediumType.image);
    final res = await Future.wait(albums.map((e) => e.listMedia()));
    final index = res.indexWhere((e) => e.album.name == 'All');
    if (index != -1) images.addAll(res[index].items);
    if (index == -1) {
      for (var e in res) {
        images.addAll(e.items);
      }
    }
    notifyListeners();
  }

  Future<void> openGallery() async {
    final res = await FilePicker.platform.pickFiles(
      allowMultiple: multiple,
      type: FileType.image,
    );
    if (res != null) {
      for (var element in res.files) {
        if (element.path != null) selectedImages.add(File(element.path!));
      }
    }
  }

  void captureImage(File file) {
    selectedImages.add(file);
  }

  Future<void> selectImage(Medium image) async {
    if (multiple) {
      final index = selectedImages
          .indexWhere((e) => e.path.split('/').last == image.filename);
      if (index != -1) {
        selectedImages.removeAt(index);
      } else {
        final file = await image.getFile();
        selectedImages.add(file);
      }
    } else {
      selectedImages.clear();
      final file = await image.getFile();
      selectedImages.add(file);
    }
    notifyListeners();
  }
}


lib/camera/camera_whatsapp.dart
import 'dart:async';
import 'dart:io';

import 'package:camera_camera/camera_camera.dart';
import 'package:flutter/gestures.dart';
import 'package:flutter/material.dart';
import 'package:flutter_sliding_up_panel/flutter_sliding_up_panel.dart';
import 'package:permission_handler/permission_handler.dart';
import 'package:photo_gallery/photo_gallery.dart';
import 'package:file_picker/file_picker.dart';
import 'package:whatsapp_camera/camera/view_image.dart';

class _WhatsAppCameraController extends ChangeNotifier {
  ///
  /// don't necessary to use this class
  /// this is the class to controller the actions
  ///
  _WhatsAppCameraController({this.multiple = true});

  /// permission to select multiple images
  ///
  /// multiple => default is true
  ///
  ///
  ///
  final bool multiple;
  final selectedImages = <File>[];
  var images = <Medium>[];

  Future<bool> handlerPermissions() async {
    final status = await Permission.storage.request();
    if (Platform.isIOS) {
      await Permission.photos.request();
      await Permission.mediaLibrary.request();
    }
    return status.isGranted;
  }

  bool imageIsSelected(String? fileName) {
    final index =
        selectedImages.indexWhere((e) => e.path.split('/').last == fileName);
    return index != -1;
  }

  _timer() {
    Timer.periodic(const Duration(seconds: 2), (t) async {
      Permission.camera.isGranted.then((value) {
        if (value) {
          getPhotosToGallery();
          t.cancel();
        }
      });
    });
  }

  Future<void> getPhotosToGallery() async {
    final permission = await handlerPermissions();
    if (permission) {
      final albums = await PhotoGallery.listAlbums(
        mediumType: MediumType.image,
      );
      final res = await Future.wait(albums.map((e) => e.listMedia()));
      final index = res.indexWhere((e) => e.album.name == 'All');
      if (index != -1) images.addAll(res[index].items);
      if (index == -1) {
        for (var e in res) {
          images.addAll(e.items);
        }
      }
      notifyListeners();
    }
  }

  Future<void> inicialize() async {
    _timer();
  }

  Future<void> openGallery() async {
    final res = await FilePicker.platform.pickFiles(
      allowMultiple: multiple,
      type: FileType.image,
    );
    if (res != null) {
      for (var element in res.files) {
        if (element.path != null) selectedImages.add(File(element.path!));
      }
    }
  }

  void captureImage(File file) {
    selectedImages.add(file);
  }

  Future<void> selectImage(Medium image) async {
    if (multiple) {
      final index = selectedImages
          .indexWhere((e) => e.path.split('/').last == image.filename);
      if (index != -1) {
        selectedImages.removeAt(index);
      } else {
        final file = await image.getFile();
        selectedImages.add(file);
      }
    } else {
      selectedImages.clear();
      final file = await image.getFile();
      selectedImages.add(file);
    }
    notifyListeners();
  }
}

class WhatsappCamera extends StatefulWidget {
  /// permission to select multiple images
  ///
  /// multiple => default is true
  ///
  ///
  ///how use:
  ///```dart
  ///List<File>? res = await Navigator.push(
  /// context,
  /// MaterialPageRoute(
  ///   builder: (context) => const WhatsappCamera()),
  ///);
  ///
  ///```
  ///
  final bool multiple;

  /// how use:
  ///```dart
  ///List<File>? res = await Navigator.push(
  /// context,
  /// MaterialPageRoute(
  ///   builder: (context) => const WhatsappCamera()),
  ///);
  ///
  ///```
  ///
  const WhatsappCamera({super.key, this.multiple = true});

  @override
  State<WhatsappCamera> createState() => _WhatsappCameraState();
}

class _WhatsappCameraState extends State<WhatsappCamera>
    with WidgetsBindingObserver {
  late _WhatsAppCameraController controller;
  final painel = SlidingUpPanelController();

  @override
  void dispose() {
    controller.dispose();
    painel.dispose();
    super.dispose();
  }

  @override
  void didChangeAppLifecycleState(AppLifecycleState state) {
    if (state == AppLifecycleState.resumed) {
      Navigator.pop(context);
    }
  }

  @override
  void initState() {
    super.initState();
    controller = _WhatsAppCameraController(multiple: widget.multiple);
    painel.addListener(() {
      if (painel.status.name == 'hidden') {
        controller.selectedImages.clear();
      }
    });
    WidgetsBinding.instance.addPostFrameCallback((_) {
      controller.inicialize();
    });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: Colors.black,
      body: Stack(
        children: [
          SizedBox(
            width: MediaQuery.of(context).size.width,
            child: CameraCamera(
              enableZoom: false,
              resolutionPreset: ResolutionPreset.high,
              onFile: (file) {
                controller.captureImage(file);
                Navigator.pop(context, controller.selectedImages);
              },
            ),
          ),
          Padding(
            padding: const EdgeInsets.only(top: 30),
            child: Row(
              mainAxisAlignment: MainAxisAlignment.spaceBetween,
              children: [
                IconButton(
                  color: Colors.white,
                  onPressed: (() => Navigator.pop(context)),
                  icon: const Icon(Icons.close),
                ),
                IconButton(
                  color: Colors.white,
                  onPressed: () async {
                    controller.openGallery().then((value) {
                      if (controller.selectedImages.isNotEmpty) {
                        Navigator.pop(context, controller.selectedImages);
                      }
                    });
                  },
                  icon: const Icon(Icons.image),
                ),
              ],
            ),
          ),
          Positioned(
            bottom: 96,
            left: 0.0,
            right: 0.0,
            child: GestureDetector(
              dragStartBehavior: DragStartBehavior.down,
              onVerticalDragStart: (details) => painel.expand(),
              child: SizedBox(
                height: 120,
                child: AnimatedBuilder(
                    animation: controller,
                    builder: (context, child) {
                      return Column(
                        children: [
                          if (controller.images.isNotEmpty)
                            const RotatedBox(
                              quarterTurns: 1,
                              child: Icon(
                                Icons.arrow_back_ios,
                                color: Colors.white,
                              ),
                            ),
                          Expanded(
                            child: ListView.builder(
                              itemCount: controller.images.length,
                              physics: const BouncingScrollPhysics(),
                              scrollDirection: Axis.horizontal,
                              itemBuilder: (context, index) {
                                return InkWell(
                                  onTap: () async {
                                    controller
                                        .selectImage(controller.images[index])
                                        .then((value) {
                                      Navigator.pop(
                                        context,
                                        controller.selectedImages,
                                      );
                                    });
                                  },
                                  child: Container(
                                    height: 100,
                                    width: 100,
                                    margin: const EdgeInsets.symmetric(
                                        horizontal: 5),
                                    decoration: BoxDecoration(
                                      image: DecorationImage(
                                        fit: BoxFit.cover,
                                        isAntiAlias: true,
                                        filterQuality: FilterQuality.high,
                                        image: ThumbnailProvider(
                                          highQuality: true,
                                          mediumId: controller.images[index].id,
                                        ),
                                      ),
                                    ),
                                  ),
                                );
                              },
                            ),
                          ),
                        ],
                      );
                    }),
              ),
            ),
          ),
          Center(
            child: SlidingUpPanelWidget(
              controlHeight: 0,
              panelController: painel,
              child: AnimatedBuilder(
                  animation: controller,
                  builder: (context, child) {
                    return _ImagesPage(
                      controller: controller,
                      close: () {
                        painel.hide();
                      },
                      done: () {
                        if (controller.selectedImages.isNotEmpty) {
                          Navigator.pop(context, controller.selectedImages);
                        } else {
                          painel.hide();
                        }
                      },
                    );
                  }),
            ),
          )
        ],
      ),
    );
  }
}

class _ImagesPage extends StatefulWidget {
  final _WhatsAppCameraController controller;

  ///
  /// close action
  /// how use:
  /// ```dart
  /// close: () {
  ///   //pop painel
  /// }
  /// ```
  ///
  final void Function()? close;

  ///
  /// done action
  /// how use:
  /// ```dart
  /// done: () {
  ///   //send images
  /// }
  /// ```
  ///
  final void Function()? done;

  ///
  ///
  /// this is thi page of swipe to up
  /// and show the images of gallery
  /// don`t is necessary your implementation by the final programmer
  ///
  ///
  const _ImagesPage({
    required this.controller,
    required this.close,
    required this.done,
  });

  @override
  State<_ImagesPage> createState() => __ImagesPageState();
}

class __ImagesPageState extends State<_ImagesPage> {
  @override
  Widget build(BuildContext context) {
    return Container(
      color: Colors.black,
      height: MediaQuery.of(context).size.height - 40,
      child: Column(
        children: [
          Container(
            height: 40,
            decoration: const BoxDecoration(
                color: Colors.white,
                borderRadius: BorderRadius.only(
                  topLeft: Radius.circular(15),
                  topRight: Radius.circular(15),
                )),
            child: Row(
              mainAxisAlignment: MainAxisAlignment.spaceBetween,
              children: [
                IconButton(
                  onPressed: widget.close?.call,
                  icon: const Icon(Icons.close),
                ),
                if (widget.controller.multiple)
                  Text(widget.controller.selectedImages.length.toString()),
                IconButton(
                  onPressed: widget.done?.call,
                  icon: const Icon(Icons.check),
                )
              ],
            ),
          ),
          Expanded(
            child: GridView.builder(
              padding: const EdgeInsets.symmetric(horizontal: 5),
              itemCount: widget.controller.images.length,
              physics: const BouncingScrollPhysics(),
              gridDelegate: SliverGridDelegateWithFixedCrossAxisCount(
                crossAxisCount: 3,
                mainAxisExtent: 140,
                crossAxisSpacing: 5,
                mainAxisSpacing: 5,
                childAspectRatio: MediaQuery.of(context).size.width /
                    (MediaQuery.of(context).size.height / 4),
              ),
              itemBuilder: (context, index) {
                return InkWell(
                  onTap: () => widget.controller
                      .selectImage(widget.controller.images[index]),
                  child: _ImageItem(
                    selected: widget.controller.imageIsSelected(
                      widget.controller.images[index].filename,
                    ),
                    image: widget.controller.images[index],
                  ),
                );
              },
            ),
          ),
        ],
      ),
    );
  }
}

class _ImageItem extends StatelessWidget {
  ///
  /// medium image
  /// is formatter usage for package: photo_gallery
  /// this package list all images of device
  ///
  final Medium image;

  ///
  /// where selected is true, apply a check in the image
  ///
  final bool selected;

  ///
  ///this widget is usage how itemBuilder of painel: _ImagesPage
  ///
  const _ImageItem({required this.image, required this.selected});

  @override
  Widget build(BuildContext context) {
    return InkWell(
      child: Stack(
        children: [
          Hero(
            tag: image.id,
            child: Container(
              decoration: BoxDecoration(
                image: DecorationImage(
                  fit: BoxFit.cover,
                  filterQuality: FilterQuality.high,
                  image: ThumbnailProvider(
                    mediumId: image.id,
                    highQuality: true,
                    height: 150,
                    width: 150,
                    mediumType: MediumType.image,
                  ),
                ),
              ),
            ),
          ),
          if (selected)
            Container(
              color: Colors.grey.withOpacity(.3),
              child: Center(
                child: Stack(
                  children: [
                    const Icon(
                      Icons.done,
                      size: 52,
                      color: Colors.white,
                    ),
                    Icon(
                      Icons.done,
                      size: 50,
                      color: Theme.of(context).primaryColor,
                    ),
                  ],
                ),
              ),
            ),
          Align(
            alignment: Alignment.topRight,
            child: IconButton(
              color: Colors.white,
              icon: const Icon(Icons.zoom_out_map_outlined),
              onPressed: () async {
                image.getFile().then((value) {
                  Navigator.of(context).push(MaterialPageRoute(
                    builder: (context) {
                      return Hero(
                        tag: image.id,
                        child: ViewImage(image: value.path),
                      );
                    },
                  ));
                });
              },
            ),
          )
        ],
      ),
    );
  }
}
